core/src/main/scala/tamer/Tamer.scala

```diff
@@ -164,51 +141,50 @@ object Tamer {
-          log.warn(
-            s"applying the following correction to $stateGroupId:\n${corrected.values.map(_.metadata.getOrElse("No Correction Metadata")).mkString("\n")}"
-          ) *>
-          adminClient.alterConsumerGroupOffsets(stateGroupId, corrected) *>
-          log.warn(s"offset of $stateGroupId corrected. Retrying initialization of state consumer to $stateTopic with group id $stateGroupId ...") *>
-          initialize
       case Die =>
-        log.error(s"consumer group $stateGroupId is already at the end of $stateTopic, manual intervention required") *>
+        log.error(s"consumer group $stateGroupId had unexpected lag for one of the $stateTopic partitions, manual intervention required") *>
           ZIO.fail(TamerError(s"Consumer group $stateGroupId stuck at end of stream"))
     }
 
-    val stateStream = stateConsumer
+    val stateStream = consumer
       .plainStream(stateKeySerde, stateValueSerde)
       .mapM {
         case CommittableRecord(record, offset) if record.key == key =>
-          log.debug(s"consumer group $stateGroupId consumed state ${record.value} from ${offset.info}") *>
-            // we got a valid state to process, invoke the handler
-            log.debug("invoking the iteration function") *> iterationFunction(record.value, queue).flatMap { newState =>
-              // now that the handler has concluded its job, we've got to commit the offset and push the new state to Kafka
-              // we should do these two operations within a transactional boundary as there is no guarantee whatsoever that
-              // both will successfully complete
-              // here we're chosing to commit first before pushing the new state so we should never lag more than 1 but
-              // we may repush the last batch when we resume (at least once semantics)
-              offset.commitOrRetry(tenTimes) <*
-                log.debug(s"consumer group $stateGroupId committed offset ${offset.info}") <*
-                stateProducer
-                  .produce(stateTopic, key, newState, stateKeySerde, stateValueSerde)
-                  .tap(rmd => log.debug(s"pushed state $newState to $rmd"))
+          producer.createTransaction.use { t =>
+            val queueDataTrans = ZQueue.bounded[Chunk[(K, V)]](requestedCapacity = 10_000).map(_.map(_.map { case (k, v) => (t, k, v) }))
+            queueDataTrans >>= { qdt =>
+              val streamOfData = ZStream.fromQueue(qdt).tap(queue.offer).runDrain
+              streamOfData <&> log.debug(s"consumer group $stateGroupId consumed state ${record.value} from ${offset.info}") *>
+                // we got a valid state to process, invoke the handler
+                log.debug(s"invoking the iteration function under $stateGroupId") *> iterationFunction(record.value, qdt).flatMap { newState =>
+                  //              (qdt.takeAll >>= queue.offerAll)
+                  UIO(()).repeatWhileM(_ => qdt.size.map(_ == 0)) *> qdt.shutdown *>
```

Comment:
This is basically
```
while (true) { if (queue.isEmpty) break; }
```
I can't find any nicer way to signal that the queue is done and the stream should be drained and closed, but I'm sure there is.